src/libNetwork/RumorManager.cpp

```diff
@@ -0,0 +1,298 @@
+    if (t == RRS::Message::Type::EMPTY_PUSH
+        || t == RRS::Message::Type::EMPTY_PULL)
+    {
+        /* Don't add it to local RumorMap because it's not the rumor itself */
+        LOG_GENERAL(DEBUG,
+                    "Received empty message of type: "
+                        << RRS::Message::s_enumKeyToString[t]);
+    }
+    else
+    {
+        auto it = m_rumorIdRumorBimap.right.find(message);
+        if (it == m_rumorIdRumorBimap.right.end())
+        {
+            recvdRumorId = ++m_rumorIdGenerator;
+            LOG_GENERAL(INFO,
+                        "New Gossip message received from: "
+                            << from << ". And new RumorId is " << recvdRumorId);
+            m_rumorIdRumorBimap.insert(
+                RumorIdRumorBimap::value_type(recvdRumorId, message));
+            toBeDispatched = true;
+        }
+        else // already received , pass it on to member for state calculations
+        {
+            recvdRumorId = it->second;
+            LOG_GENERAL(INFO,
+                        "Old Gossip message from "
+                            << from << ". And old RumorId is " << recvdRumorId);
+        }
+    }
+
+    RRS::Message recvMsg(t, recvdRumorId, round);
+
+    int peerId = p->second;
+    std::pair<int, std::vector<RRS::Message>> pullMsgs
+        = m_rumorHolder->receivedMessage(recvMsg, peerId);
+
+    // Get the corresponding Peer to which to send Pull Messages if any.
+    auto l = m_peerIdPeerBimap.left.find(pullMsgs.first);
+    if (l != m_peerIdPeerBimap.left.end())
+    {
+        LOG_GENERAL(DEBUG,
+                    "Sending " << pullMsgs.second.size() << " PULL Messages");
+        SendMessages(l->second, pullMsgs.second);
+    }
+
+    return toBeDispatched;
+}
+
+void RumorManager::SendMessages(const Peer& toPeer,
+                                const std::vector<RRS::Message>& messages)
+{
+    for (auto& k : messages)
+    {
+        // Add round and type to outgoing message
+        RawBytes cmd = {(unsigned char)k.type()};
+        unsigned int cur_offset = RRSMessageOffset::R_AGE;
+
+        Serializable::SetNumber<uint32_t>(cmd, cur_offset, k.age(),
+                                          sizeof(uint32_t));
+
+        cur_offset += sizeof(uint32_t);
+
+        Serializable::SetNumber<uint32_t>(
+            cmd, cur_offset, m_selfPeer.m_listenPortHost, sizeof(uint32_t));
+
+        // Get the raw messages based on rumor ids.
+        auto m = m_rumorIdRumorBimap.left.find(k.rumorId());
+        if (m != m_rumorIdRumorBimap.left.end())
+        {
+            // Add raw message to outgoing message
+            cmd.insert(cmd.end(), m->second.begin(), m->second.end());
+            LOG_GENERAL(DEBUG,
+                        "Sending Non Empty - Gossip Message - "
+                            << k << " To Peer : " << toPeer);
+        }
+
+        // Send the message to peer .
+        P2PComm::GetInstance().SendMessageNoQueue(toPeer, cmd,
```

Comment:
I would suggest not to use the `NoQueue` version anywhere in `RumorManager`. Historically, this function was added only for test code that didn't necessarily have all the classes instantiated, and where we basically just want to send a message. In our actual code, using the `NoQueue` version would skip using the sending queue, which is supposed to manage all the outgoing messages.

Reply:
After discussion, will keep this way for now